option java_package = "de.arago.autopilot.client.zmq.protobuf";

message EmptyMessage {
}

message StringMessage {
    required string value = 1;
}

message BooleanMessage {
    required bool value = 2;
}

message Int64Message {
    required int64 value = 3;
}

message StringResponse {
    required string value = 4;
    required bool success = 5;
}

message StringListResponse {
    repeated string list = 6;
    required bool success = 7;
}

message TwoIDsMessage {
    required string id1 = 1;
    required string id2 = 2;
}

message GetAttributeRequest {
    required string nodeid = 1;
    required string parentuid = 2;
    required string attr = 3;
}

message SetAttributeRequest {
    required string nodeid = 1;
    required string parentuid = 2;
    required string attr = 3;
    required string value = 4;
}

message DelAttributeRequest {
    required string issueid = 1;
    required string parentuid = 2;
    required string attr = 3;
}

message BaseObjectMessage {
    required string name = 1;
    required string id = 2;
    required string type = 3;
    required int64 ctime = 4;
    required int64 mtime = 5;
}

message KeyValueMessage {
    required string key = 1;
    required string value = 2;
}

message BaseDataMessage {
    required BaseObjectMessage base = 1;
    required string nodeid = 2;
    optional string parentuid = 3;
    repeated KeyValueMessage attributes = 9;
    repeated BaseDataMessage sub = 10;
}

message StringOrErrorResponse {
    required string value = 3;
    required string message = 4;
    required bool success = 5;
}
